web/web.go

```diff
@@ -402,6 +404,15 @@ func New(logger log.Logger, o *Options) *Handler {
 		router.Get("/user/*filepath", route.FileServe(o.UserAssetsPath))
 	}
 
+	langserver, err := langserver.CreateHandler(context.Background(), fmt.Sprint("http://", o.ListenAddress))
```

Comment:
Whats the right way to deal with an error raised here?

There are two cases which could cause an error here:

* A invalid URL argument.
* A bug in the language server.


(I'm aware the `fmt.Println` statement below isn't proper error handling)

Reply:
In either case, Prometheus should fail to start. It sounds like you're not handling external url correctly here though.

Reply:
Yeah, I guess we'd always want to connect to localhost instead of the external URL's hostname though (which might not even be reachable from the Prometheus machine), but we'd still have to append the route prefix.

Even better would be if we could avoid the network completely for these internal calls. Like if the langserver handler could take a Prometheus API object or HTTP handler, instead of a Prometheus URL to connect to. That's just a very rough idea though, no idea how doable that is :)

Reply:
I think we should indeed avoid http requests here.

Reply:
Yeah, this shouldn't be hitting the network stack at all. We've had enough problems from that in the past.

Reply:
https://github.com/prometheus-community/promql-langserver/issues/117

Reply:
The language server now has been updated in a way that allows not using the network stack (Thanks to @Nexucis !). This PR will be eventually updated accordingly.

Reply:
You are welcome :)

Just to complete a bit what @slrtbfs said, I created an interface that handle the request to Prometheus:
https://github.com/prometheus-community/promql-langserver/blob/11052ee3baf9ed3f5ddfc1d604ba819bbf21377b/prometheus/client.go#L88-L104

The method createHandler is taking in parameter this interface. 
What is remain now is to create a client that implements this interface and that won't use the network.

But the main question de have today with @slrbtfs is where to put this implementation. ( Talk is here https://github.com/prometheus-community/promql-langserver/issues/134) For me it makes sense to do it on Prometheus side and to not have it in the langserver side.

Do you think the same ?

Reply:
@Nexucis Others who are deeper into this may speak up and disagree, but from my current understanding I would also think that a Prometheus-server-specific (non-networked) implementation of that metadata client would fit best into the Prometheus repo itself.